src/lib/rule2F_test.ts

```diff
@@ -0,0 +1,215 @@
+import {html, render} from 'lit-html';
+import {rule2F, allowsNameFromContent_TEST} from './rule2F';
+import {Context, getDefaultContext} from './context';
+
+describe('The function for rule 2F', () => {
+  let container: HTMLElement;
+  let context: Context;
+  beforeEach(() => {
+    container = document.createElement('div');
+    document.body.appendChild(container);
+    context = getDefaultContext();
+  });
+
+  afterEach(() => {
+    document.body.removeChild(container);
+  });
+
+  it('returns text content of role="button" nodes', () => {
+    render(
+      html`
+        <div id="foo" role="button">
+          Hello world
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    expect(rule2F(elem!, context)).toBe('Hello world');
+  });
+
+  it('returns text content of subtree if node label element', () => {
+    render(
+      html`
+        <label id="foo">
+          <div>
+            Hello
+            <div>
+              <div>
+                world
+              </div>
+            </div>
+          </div>
+          <div>!</div>
+        </label>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Hello world !');
+  });
+
+  it('considers CSS generated text content', () => {
+    render(
+      html`
+        <div id="foo">
+          world
+          <div>
+            <style>
+              #foo:before {
+                content: 'Hello';
+              }
+              #foo:after {
+                content: '!';
+              }
+            </style>
+          </div>
+        </div>
+      `,
+      container
+    );
+    const elem = document.getElementById('foo');
+    context.isLabelReference = true;
+    expect(rule2F(elem!, context)).toBe('Helloworld!');
+  });
+
+  it('considers entire subtree for descendants of label reference', () => {
+    render(
+      html`
+        <div id="foo">
+          Hello
+          <div aria-labelledby="bar"></div>
+        </div>
+        <div id="bar">
+          world
+          <div aria-labelledby="foo"></div>
```

Comment:
This is also testing the infinite recursion protection right? I'd split this out into another test `it('doesn't visit the same node twice during recursion'`